examples/github-deploy.yml

```
@@ -0,0 +1,78 @@
+name: Build and upload to PyPI
+
+on: [push, pull_request]
+# alternatively, to publish when a (published) GitHub Release is created, use the following:
+# on:
+#   release:
+#     types:
+#       - published
+
+jobs:
+  build_wheels:
+    name: Build wheels on ${{ matrix.os }}
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        os: [ubuntu-18.04, windows-latest, macos-latest]
+
+    steps:
+      - uses: actions/checkout@v2
+
+      - uses: actions/setup-python@v2
+        name: Install Python
+        with:
+          python-version: '3.7'
+
+      - name: Install cibuildwheel
+        run: |
+          python -m pip install cibuildwheel==1.4.2
+
+      - name: Install Visual C++ for Python 2.7
+        if: startsWith(matrix.os, 'windows')
+        run: |
+          choco install vcpython27 -f -y
+
+      - name: Build wheels
+        run: |
+          python -m cibuildwheel --output-dir wheelhouse
+
+      - uses: actions/upload-artifact@v2
+        with:
+          path: ./wheelhouse/*.whl
+
+  build_sdist:
+    name: Build source distribution
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+
+      - uses: actions/setup-python@v2
+        name: Install Python
+        with:
+          python-version: '3.7'
+
+      - name: Build sdist
+        run: python setup.py sdist
```

Instead of having a setuptools lock-in, this could use PEP 517 interface:
```suggestion
        run: |
          python -m pip install pep517
          python -m pep517.build --source --out-dir dist/ .
```

---

Could we replace pip wheel builds with pep517? (does it work for classic setuptools packages, is the PEP 518 environment setup fine, would it fix the extra installed wheels in the wheelhouse issue)

I've seen it, and know that pip vendors it internally, but not sure how well it works standalone.

---

For just the minimal example, this feels a lot heavier though. Won't users also find the setuptools approach sooner? E.g., here: https://packaging.python.org/tutorials/packaging-projects/#packaging-your-project

And I suppose that once someone is not using basic `setuptools`, it's obvious enough this line needs to be replaced?

Or is this the new default way to build a source distribution that's meant to be spread and replace `setup.py sdist`?

---

I didn't think pep517 was really intended for users, at least yet, but as a backend for other tools, like pip. It's used internally to some extent by pip, I believe, and for the build command, there still is this note on the package readme:

> This ‘build’ module should be considered experimental while the PyPA decides on the best place for this functionality.

I don't think it's work putting into an example quite yet.

---

Currently, setuptools is the only system I know of that produces binaries, and cibuildwheel builds binaries, so lock-in is not yet so important to avoid. ;)

---

Thanks, @henryiii. Is that OK, @webknjaz?
We can still easily change this later, once the Python docs update best practices, etc.